### ISSUE TITLE
Error importing SQLite DB

### ISSUE THREAD

**Issue by TheOneWithTheBraid** [spantaleev/matrix-docker-ansible-deploy#775]
When importing the SQLite DB from a previous synapse installation, I am told I was trying to write a read-only database.

Does anyone have a clue what might cause this issue?

I retried on a clean new ansible playbook. The provided homeserver.db is actually propper.

<details>

<summary> Error message </summary>

```
TASK [matrix-postgres : Import SQLite database into Postgres] *********************************************************************************************************************************************************************************
fatal: [matrix.activism.international]: FAILED! => {"changed": true, "cmd": ["docker", "run", "--rm", "--name=matrix-synapse-migrate", "--log-driver=none", "--user=997:1002", "--cap-drop=ALL", "--network=matrix", "--entrypoint=python", "--mount", "type=bind,src=/matrix/synapse/config,dst=/data", "--mount", "type=bind,src=/matrix/synapse/config,dst=/matrix-media-store-parent/media-store", "--mount", "type=bind,src=/home/lanna/synapse-data/homeserver.db,dst=/homeserver.db", "docker.io/matrixdotorg/synapse:v1.24.0", "/usr/local/bin/synapse_port_db", "--sqlite-database", "/homeserver.db", "--postgres-config", "/data/homeserver.yaml"], "delta": "0:00:04.578427", "end": "2021-01-07 12:39:37.697483", "msg": "non-zero return code", "rc": 5, "start": "2021-01-07 12:39:33.119056", "stderr": "2021-01-07 12:39:36,856 - synapse.config.key - 145 - WARNING - This server is configured to use 'matrix.org' as its trusted key server via the\n'trusted_key_servers' config option. 'matrix.org' is a good choice for a key\nserver since it is long-lived, stable and trusted. However, some admins may\nwish to use another server for this purpose.\n\nTo suppress this warning and continue using 'matrix.org', admins should set\n'suppress_key_server_warning' to 'true' in homeserver.yaml.\n--------------------------------------------------------------------------------\n2021-01-07 12:39:36,888 - synapse.storage.prepare_database - 102 - INFO - ['main', 'state']: Checking existing schema version\n2021-01-07 12:39:36,893 - synapse.storage.prepare_database - 107 - INFO - ['main', 'state']: Existing schema is 58 (+30 deltas)\n2021-01-07 12:39:36,894 - synapse.storage.databases.main - 408 - INFO - Checking database for consistency with configuration...\n2021-01-07 12:39:36,895 - synapse.storage.prepare_database - 377 - INFO - Applying schema deltas for v58\n2021-01-07 12:39:36,896 - synapse.storage.prepare_database - 464 - INFO - Applying schema 58/24drop_event_json_index.sql\n2021-01-07 12:39:36,897 - synapse_port_db - 704 - ERROR - \nTraceback (most recent call last):\n  File \"/usr/local/bin/synapse_port_db\", line 562, in run\n    self.sqlite_store = self.build_db_store(\n  File \"/usr/local/bin/synapse_port_db\", line 528, in build_db_store\n    prepare_database(db_conn, engine, config=self.hs_config)\n  File \"/usr/local/lib/python3.8/site-packages/synapse/storage/prepare_database.py\", line 128, in prepare_database\n    _upgrade_existing_database(\n  File \"/usr/local/lib/python3.8/site-packages/synapse/storage/prepare_database.py\", line 465, in _upgrade_existing_database\n    executescript(cur, absolute_path)\n  File \"/usr/local/lib/python3.8/site-packages/synapse/storage/prepare_database.py\", line 599, in executescript\n    execute_statements_from_stream(txn, f)\n  File \"/usr/local/lib/python3.8/site-packages/synapse/storage/prepare_database.py\", line 604, in execute_statements_from_stream\n    cur.execute(statement)\n  File \"/usr/local/lib/python3.8/site-packages/synapse/storage/database.py\", line 271, in execute\n    self._do_execute(self.txn.execute, sql, *args)\n  File \"/usr/local/lib/python3.8/site-packages/synapse/storage/database.py\", line 297, in _do_execute\n    return func(sql, *args)\nsqlite3.OperationalError: attempt to write a readonly database\nTraceback (most recent call last):\n  File \"/usr/local/bin/synapse_port_db\", line 562, in run\n    self.sqlite_store = self.build_db_store(\n  File \"/usr/local/bin/synapse_port_db\", line 528, in build_db_store\n    prepare_database(db_conn, engine, config=self.hs_config)\n  File \"/usr/local/lib/python3.8/site-packages/synapse/storage/prepare_database.py\", line 128, in prepare_database\n    _upgrade_existing_database(\n  File \"/usr/local/lib/python3.8/site-packages/synapse/storage/prepare_database.py\", line 465, in _upgrade_existing_database\n    executescript(cur, absolute_path)\n  File \"/usr/local/lib/python3.8/site-packages/synapse/storage/prepare_database.py\", line 599, in executescript\n    execute_statements_from_stream(txn, f)\n  File \"/usr/local/lib/python3.8/site-packages/synapse/storage/prepare_database.py\", line 604, in execute_statements_from_stream\n    cur.execute(statement)\n  File \"/usr/local/lib/python3.8/site-packages/synapse/storage/database.py\", line 271, in execute\n    self._do_execute(self.txn.execute, sql, *args)\n  File \"/usr/local/lib/python3.8/site-packages/synapse/storage/database.py\", line 297, in _do_execute\n    return func(sql, *args)\nsqlite3.OperationalError: attempt to write a readonly database\nattempt to write a readonly database", "stderr_lines": ["2021-01-07 12:39:36,856 - synapse.config.key - 145 - WARNING - This server is configured to use 'matrix.org' as its trusted key server via the", "'trusted_key_servers' config option. 'matrix.org' is a good choice for a key", "server since it is long-lived, stable and trusted. However, some admins may", "wish to use another server for this purpose.", "", "To suppress this warning and continue using 'matrix.org', admins should set", "'suppress_key_server_warning' to 'true' in homeserver.yaml.", "--------------------------------------------------------------------------------", "2021-01-07 12:39:36,888 - synapse.storage.prepare_database - 102 - INFO - ['main', 'state']: Checking existing schema version", "2021-01-07 12:39:36,893 - synapse.storage.prepare_database - 107 - INFO - ['main', 'state']: Existing schema is 58 (+30 deltas)", "2021-01-07 12:39:36,894 - synapse.storage.databases.main - 408 - INFO - Checking database for consistency with configuration...", "2021-01-07 12:39:36,895 - synapse.storage.prepare_database - 377 - INFO - Applying schema deltas for v58", "2021-01-07 12:39:36,896 - synapse.storage.prepare_database - 464 - INFO - Applying schema 58/24drop_event_json_index.sql", "2021-01-07 12:39:36,897 - synapse_port_db - 704 - ERROR - ", "Traceback (most recent call last):", "  File \"/usr/local/bin/synapse_port_db\", line 562, in run", "    self.sqlite_store = self.build_db_store(", "  File \"/usr/local/bin/synapse_port_db\", line 528, in build_db_store", "    prepare_database(db_conn, engine, config=self.hs_config)", "  File \"/usr/local/lib/python3.8/site-packages/synapse/storage/prepare_database.py\", line 128, in prepare_database", "    _upgrade_existing_database(", "  File \"/usr/local/lib/python3.8/site-packages/synapse/storage/prepare_database.py\", line 465, in _upgrade_existing_database", "    executescript(cur, absolute_path)", "  File \"/usr/local/lib/python3.8/site-packages/synapse/storage/prepare_database.py\", line 599, in executescript", "    execute_statements_from_stream(txn, f)", "  File \"/usr/local/lib/python3.8/site-packages/synapse/storage/prepare_database.py\", line 604, in execute_statements_from_stream", "    cur.execute(statement)", "  File \"/usr/local/lib/python3.8/site-packages/synapse/storage/database.py\", line 271, in execute", "    self._do_execute(self.txn.execute, sql, *args)", "  File \"/usr/local/lib/python3.8/site-packages/synapse/storage/database.py\", line 297, in _do_execute", "    return func(sql, *args)", "sqlite3.OperationalError: attempt to write a readonly database", "Traceback (most recent call last):", "  File \"/usr/local/bin/synapse_port_db\", line 562, in run", "    self.sqlite_store = self.build_db_store(", "  File \"/usr/local/bin/synapse_port_db\", line 528, in build_db_store", "    prepare_database(db_conn, engine, config=self.hs_config)", "  File \"/usr/local/lib/python3.8/site-packages/synapse/storage/prepare_database.py\", line 128, in prepare_database", "    _upgrade_existing_database(", "  File \"/usr/local/lib/python3.8/site-packages/synapse/storage/prepare_database.py\", line 465, in _upgrade_existing_database", "    executescript(cur, absolute_path)", "  File \"/usr/local/lib/python3.8/site-packages/synapse/storage/prepare_database.py\", line 599, in executescript", "    execute_statements_from_stream(txn, f)", "  File \"/usr/local/lib/python3.8/site-packages/synapse/storage/prepare_database.py\", line 604, in execute_statements_from_stream", "    cur.execute(statement)", "  File \"/usr/local/lib/python3.8/site-packages/synapse/storage/database.py\", line 271, in execute", "    self._do_execute(self.txn.execute, sql, *args)", "  File \"/usr/local/lib/python3.8/site-packages/synapse/storage/database.py\", line 297, in _do_execute", "    return func(sql, *args)", "sqlite3.OperationalError: attempt to write a readonly database", "attempt to write a readonly database"], "stdout": "Preparing sqlite3...", "stdout_lines": ["Preparing sqlite3..."]}
```

</details>

**Comment by spantaleev** [comment]
Make sure the SQLite `.db` file is writable.

I'm guessing it's from an older version of Synapse and it tries to run database migrations against it, before actually importing it into Postgres.

`chown matrix:matrix /path/to/file.db` and `chmod o+w /path/to/file.db`

------------

I guess the playbook may try to do this automatically, especially once you confirm that changing file ownership helps.

**Comment by TheOneWithTheBraid** [comment]
Thanks. Unfortunately, it didn't solve my issue. It still throws exactly the same error.

**Comment by TheOneWithTheBraid** [comment]
Oh, spontaneously it seems to be due to wrong GID and UID. Does the synapse container use `991:991` as default user of synapse?

**Comment by spantaleev** [comment]
Importing happens like this: https://github.com/spantaleev/matrix-docker-ansible-deploy/blob/1dec01095ac87676cd5a38041059247b5872d18e/roles/matrix-postgres/tasks/import_synapse_sqlite_db.yml#L72-L86

It uses the user id and group id for the `matrix` user and `matrix` group. This user and group are automatically created and determined, so it's not necessarily `991`.

Old installations of the playbook were hardcoding them to `991`, so some of us would have them be like that. Otherwise, it's not the case.

It's why I said you should `chown` to `matrix:matrix`.

**Comment by TheOneWithTheBraid** [comment]
Thank you so much. Now, it worked.

I did the following:
- update the legacy synapse installation and restart the service
- this updates the DB to the most current scheme
- import it into ansible using the above mentioned preferences

**Comment by Lyokovic** [comment]
Commenting on this issue because I faced the same one and it seems that there is a real issue.
I tried to import my sqlite3 DB, and it did not work because the schema upgrade process failed because of a read-only database.
The database file owner is the `matrix` user, but it did not help.

I found that copying the database to the synapse config directory and changing the `docker run ... synapse_port_db --sqlite-database` command argument to the database file in the config directory did remove the read-only issue.

I am wondering if the fact that the file itself is bind-mounted in the container (and not a directory containing the file) makes it read-only.
